src/polyswarmd/websockets/messages.py

```diff
@@ -1,386 +1,493 @@
-        "Extract the fields indicated in _extraction_schema from the event log message"
-        return copy_with_schema(cls._extraction_schema, source)
+    def __init_subclass__(cls):
+        super().__init_subclass__()
+        cls.contract_event_name = cls.__name__
 
     @classmethod
-    def contract_event_name(cls) -> str:
-        "The event name used by web3 (e.g 'Transfer' or 'FeesUpdated')"
-        return str(cls.__name__)
-
-    def __repr__(self):
-        return f'<EventLogMessage contract_event_name={self.contract_event_name()}>'
+    def extract(cls: Any, instance: Mapping) -> E:
+        "Extract the fields indicated in schema from the event log message"
+        return cls.schema.extract(instance)
 
 
 # Commonly used schema properties
-uint256 = {'type': 'integer'}
-guid = {'type': 'string', 'format': 'uuid', '$#convert': True}
-bounty_guid = {**guid, '$#from': 'bountyGuid'}
-ethereum_address = {'type': 'string', 'format': 'ethaddr'}
-
-# The fetch routine uses format to split into a bitvector, e.g format(16, "0>10b") => '0000010000'
-boolvector = {
-    'type': 'array',
-    'items': 'boolean',
-    '$#fetch': lambda e, k, *args: [b != '0' for b in format(int(e[k]), f"0>{e.numArtifacts}b")]
-}
+uint256: SchemaDef = {'type': 'integer'}
+guid: SchemaDef = {'type': 'string', 'format': 'uuid'}
+bounty_guid: SchemaDef = cast(SchemaDef, {**guid, 'srckey': 'bountyGuid'})
+ethereum_address: SchemaDef = {'format': 'ethaddr', 'type': 'string'}
 
-# The functions below are commented out because they depend on configuration being loaded from PolyswarmD.
-from requests_futures.sessions import FuturesSession
 
-session = FuturesSession(adapter_kwargs={'max_retries': 3})
+def int_to_bool_list(i):
+    s = format(i, 'b')
+    return [x == '1' for x in s[::-1]]
 
-config = None
-artifact_client = None
-redis = None
 
-from functools import lru_cache
+def safe_int_to_bool_list(num, max):
+    if int(num) == 0:
+        return [False] * int(max)
+    else:
+        converted = int_to_bool_list(num)
+        return converted + [False] * (max - len(converted))
 
 
-# XXX this is about as hacky as it gets. We should extract configuration loading out.
-@lru_cache(15)
-def fetch_metadata(uri: str, validate):
-    from polyswarmd.bounties import substitute_metadata
-    global config
-    global redis
-    global artifact_client
-    if config is None:
-        from polyswarmd import app
-        config = app.config
-        artifact_client = config['POLYSWARMD'].artifact_client
-        redis = config['POLYSWARMD'].redis
+boolvector: SchemaDef = {
+    'type': 'array',
+    'items': 'boolean',
+    'srckey': lambda k, e: safe_int_to_bool_list(e[k], e['numArtifacts'])
+}
 
-    return substitute_metadata(uri, artifact_client, session, validate=validate, redis=redis)
+# partially applied `substitute_metadata' with AI, redis & session prefilled.
+_substitute_metadata: Optional[Callable[[str, bool], Any]] = None
```

Comment:
Should this be a class property instead of a module global variable? Create a class for metadata?

Reply:
Yeah, that's absolutely right. Making the change now.